### PR TITLE
Edge is no longer a Pair

### PR DESCRIPTION
**Pull request by sbromberger**
This is the first step in abstracting edges to handle various other types. Things I'm thinking about:

1) Edges vs Arcs
2) Abstractions of vertex types

### CODE REVIEW

**Comment by sbromberger** [comment]
Sigh - a transient CI error.

**Comment by IssamT** [comment]
1) Having a specific type for Edge that is not Pair is nice, but why not keep the same structure?

```julia
struct Pair{A,B}
    first::A
    second::B
end

const => = Pair
```

See also:
https://github.com/JuliaLang/julia/issues/259

2) When it comes to differentiating Edge from Arc, this is a nice abstraction, but it does bring a lot of complexity implementation wise. When I first started using LightGraphs I didn't like edges and arcs being of the same type. But when I started implementing complex algorithms, I was glad that edges and arcs were both simple pairs.

**Comment by sbromberger** [comment]
We have a chance to call the structure elements by their proper descriptive names. "First" and "second" mean something to a pair, but not to an edge. We provide accessors to these elements so nobody should be using them for precisely this reason. 

Arcs vs Edges: not sure how this will play out, but if we do it both will be derived from some common abstract class.

**Comment by sbromberger** [comment]
Oh, if you were talking about using struts: we could look into that but he goal is full immutability with type parameterization and no overhead. I'm not quite sure what struct buys us over immutable. Can you elaborate?

**Comment by IssamT** [comment]
Yes exactly. I meant to change only what we want to have different from Pair (including naming the attributes `src` and `dst`) but nothing more.

Honestly I am not sure about the exact implementation choices/details in this file.

https://github.com/JuliaLang/julia/blob/master/base/pair.jl

But I just feel more comfortable using types provided by the base.

I agree that we should have a type `Edge` so that we can do SOME things in a different way than it is done with `Pair` but we should only change those things and duplicate the rest of the code in the above file

For example if we don't have the equivalent of those methods on our type `Edge`

```julia
start(p::Pair) = 1
done(p::Pair, i) = i>2
next(p::Pair, i) = (getfield(p,i), i+1)
````

we can't use

```julia
for v in e
```

and we'll need to write something like

```julia
for v in [src(e), dst(e)]
```

**Comment by sbromberger** [comment]
"struct fields can only be bits types, other struct types, or refs"

That seems to preclude strings or other artificial data structures as elements. It's ok for SimpleGraphs but it's probably not a limitation we want to burden others with. 

Also, it appears that structs are mutable. This may actually be beneficial for memory allocations but I'd rather stick with something that I know can't be changed.

File: src/core.jl [hunk 1]
```diff
@@ -4,18 +4,24 @@ abstract AbstractPathState
...
+show(io::IO, e::Edge) = print(io, "Edge $(e.src) => $(e.dst)")
 
 """Return source of an edge."""
-src(e::Edge) = e.first
+src(e::Edge) = e.src
 """Return destination of an edge."""
-dst(e::Edge) = e.second
+dst(e::Edge) = e.dst
+
+srcdst(e::Edge) = (src(e), dst(e))
```
Comment (jpfairbanks) [inline src/core.jl]: Could this be a method of `convert(::Tuple{Int,Int}, e::Edge)` rather than a new function?

**Comment by jpfairbanks** [comment]
> we can't use
> 
> for v in e
> and we'll need to write something like
> 
> for v in (src(e), dst(e))

I am not sure that I like `for v in e` because it relies on an implicit ordering of src then dst.
Do you have an example where you use `for v in e` in  a meaningful way?

**Comment by sbromberger** [comment]
`for v in e` is not a great idea for lots of reasons. However, we do have a new accessor for this: `srcdst(e)` will return a pair of vertices.

The reason we don't want to treat edges as sequences of vertices is because when we start doing things like weighted edges, will the expectation be that we ignore the weight? `for v in e` when e consists of src, dst, and weight doesn't make intuitive sense.

Also, treating an edge like a sequence is dangerously close to accessing the fields directly. We need to provide accessors for this if it's important (the main reason I did `srcdst` is because `(i, j) = e` is fairly common, though bad practice. It's now `(i, j) = srcdst(e)`).

I do need to export `srcdst`, though.

File: src/core.jl (hunk 1 above)
Comment (sbromberger) [inline src/core.jl]: How would you use it?

**Comment by IssamT** [comment]
@jpfairbanks 
JuMP for example accepts any iterable as an index set. if un edge is not iterable i can not use it as an index set

@sbromberger 
Yes I saw the issue about adding weights to edges but what if there are two weights for a given algorithm (for instance the cost and the capactiy) ? I thought that the idea of LightGraphs was to keep the data structure as light as possible and handle additional data (not required by all the algorithms) outside the structure, with Networks.jl being a convenient solution for that... 

I am fine with srcdst solution although I wish I could avoid it.

How about the other methods of `Pair` ? I do find them convenient at times.

My issues here are not very strong (mostly related to convenience) and I am fine with any solution you go for. You probably work more often with LightGraphs and know what is better :)

**Comment by sbromberger** [comment]
> I thought that the idea of LightGraphs was to keep the data structure as light as possible and handle additional data (not required by all the algorithms) outside the structure

We're experimenting here. See #526.

**Comment by sbromberger** [comment]
Paging Dr. @jpfairbanks... @jpfairbanks to the reviewing platform, please :)

File: src/core.jl [hunk 2]
```diff
@@ -4,18 +4,26 @@ abstract AbstractPathState
 _insert_and_dedup!(v::Vector{Int}, x::Int) = isempty(splice!(v, searchsorted(v,x), x))
 
 """A type representing a single edge between two vertices of a graph."""
-typealias Edge Pair{Int,Int}
+abstract AbstractEdge
+immutable Edge <: AbstractEdge
```
Comment (jpfairbanks) [inline src/core.jl]: Good that this is immutable, although for 0.6 this will become `immutable struct`

**Comment by sbromberger** [comment]
Benchmarks (I pre-ran them before I realized you had duplicated): 

pair edges (existing): https://gist.github.com/sbromberger/63437d8f13bfa84ca9a95cd34cc4fbca
```
julia> @time fille(10)
  0.000008 seconds (5 allocations: 400 bytes)
julia> @time fillp(10)
  0.000005 seconds (5 allocations: 400 bytes)
julia> @time a = fille(n)
  0.000248 seconds (6 allocations: 156.484 KB)
julia> @time b = fillp(n)
  0.000334 seconds (6 allocations: 156.484 KB)
julia> @time tsum(a)
  0.000020 seconds (5 allocations: 176 bytes)
julia> @time tsum(b)
  0.000027 seconds (5 allocations: 176 bytes)
```

new edges: https://gist.github.com/sbromberger/81b9a8c7f40962eaf245f4ab1b75d1c5
```
julia> @time fille(10)
  0.000003 seconds (5 allocations: 400 bytes)
julia> @time fillp(10)
  0.000004 seconds (5 allocations: 400 bytes)
julia> @time a = fille(n)
  0.000095 seconds (6 allocations: 156.484 KB)
julia> @time b = fillp(n)
  0.000115 seconds (6 allocations: 156.484 KB)
julia> @time tsum(a)
  0.000067 seconds (5 allocations: 176 bytes)
julia> @time tsum(b)
  0.000020 seconds (5 allocations: 176 bytes)
```

**Pull request merged**